### PR TITLE
feat: decoding candidates for unverified contracts

### PR DESCRIPTION
**Pull request by zachdaniel**
## Motivation

* In order to show what function we suspect was called, we need to store and then list any potential contract ABI matches from other verified smart contracts. This will give users as much information as possible about unverified contract.

### Enhancements
* Store any new method identifiers we receive.
* List them as possibilities on transactions for unverified transactions.

screen shot 2019-02-27 at 11 36 02 am

### CODE REVIEW

**Comment by acravenho** [comment]
I'm not able to get this to work correctly. I deployed and verified a simple token here: https://blockscout.com/poa/core/address/0x8deb2e426835067f48ac034958cdd9daa38bd0a2/contracts

If you want to replicate this locally the only thing extra you might need is the constructor arguments:

```
00000000000000000000000057da3f24d81186b344649fe92b120ad45100d749
```

This contract has a simple function 

```
{"name": "transfer", "type": "function", "inputs": [{"name": "_to", "type": "address"}, {"name": "_value", "type": "uint256"}], "outputs": [{"name": "", "type": "bool"}], "payable": false, "constant": false, "stateMutability": "nonpayable"}
```

which is used by most other token transfers. They have the same 4byte signature.

I viewed a transaction with the same 4byte signature but there was no decoding of this input.

0x9f6079797ff552b661944d2fbaaa7b394b872b6d06254810a8fe2667e2e833e8

**Comment by acravenho** [comment]
It also used an identifier of `-145924998`. I'm not sure where this comes from.

**Comment by zachdaniel** [comment]
I'll take a look now.

**Comment by zachdaniel** [comment]
The error can be shown just w/in the ex_abi library in a console like so:

```
selector = %ABI.FunctionSelector{
  function: "transfer",
  input_names: ["_to", "_value"],
  inputs_indexed: nil,
  method_id: <<169, 5, 156, 187>>,
  returns: [:bool],
  type: :function,
  types: [:address, {:uint, 256}]
}

input = <<159, 96, 121, 121, 127, 245, 82, 182, 97, 148, 77, 47, 186, 170, 123, 57, 75, 135, 43, 109, 6, 37, 72, 16, 168, 254, 38, 103, 226, 232, 51, 232>>

ABI.decode(selector, input)
```

@ayrat555 would you mind taking a look also if you get the chance?

File: apps/block_scout_web/lib/block_scout_web/templates/transaction/_decoded_input.html.eex [hunk 1]
```diff
@@ -3,67 +3,34 @@
     <h1 class="card-title"><%= gettext "Input" %> </h1>
     <!-- Input -->
     <%= case @decoded_input_data do %>
-    <% {:error, :contract_not_verified} -> %>
-    <div class="alert alert-info">
-      <%= gettext "To see decoded input data, the contract must be verified." %>
+    <% {:error, :contract_not_verified, candidates} -> %>
+    <div class="alert alert-danger">
```
[the remark this answers is not in the document]
Comment (zachdaniel) [reply apps/block_scout_web/lib/block_scout_web/templates/transaction/_decoded_input.html.eex]: Sounds good!

[the remark this answers is not in the document]
Comment (zachdaniel) [reply apps/block_scout_web/lib/block_scout_web/templates/transaction/_decoded_input.html.eex]: Done.

**Comment by vbaranov** [comment]
>But I do not insist on these UI changes.

Oh, as I see we already use this UI for displaying of decoded inputs for verified contracts. Thus, forget my comment about UI 😃.

File: apps/explorer/lib/explorer/chain/method_identifier.ex [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  The first four bytes of the [KECCAK-256](https://en.wikipedia.org/wiki/SHA-3) hash of a contract method or event.
+
+  Represented in the database as a 4 byte integer, decodes into a 4 byte bitstring
+  """
+
+  @behaviour Ecto.Type
+
+  @type t :: binary
+
+  defguard one_byte(value) when 0 <= value and value <= 255
```
Comment (ayrat555) [inline apps/explorer/lib/explorer/chain/method_identifier.ex]: is this guard unused?

File: apps/explorer/lib/explorer/chain/transaction.ex [hunk 3]
```diff
@@ -466,9 +468,40 @@ defmodule Explorer.Chain.Transaction do
   def decoded_input_data(%__MODULE__{to_address: nil}), do: {:error, :no_to_address}
   def decoded_input_data(%__MODULE__{input: %{bytes: bytes}}) when bytes in [nil, <<>>], do: {:error, :no_input_data}
   def decoded_input_data(%__MODULE__{to_address: %{contract_code: nil}}), do: {:error, :not_a_contract_call}
-  def decoded_input_data(%__MODULE__{to_address: %{smart_contract: nil}}), do: {:error, :contract_not_verified}
+
+  def decoded_input_data(%__MODULE__{
```
Comment (ayrat555) [inline apps/explorer/lib/explorer/chain/transaction.ex]: Can we add a test for this case?

File: apps/explorer/lib/explorer/chain/method_identifier.ex (hunk 2 above)
Comment (zachdaniel) [reply apps/explorer/lib/explorer/chain/method_identifier.ex]: Yeah, it is unused. I'll remove it.

File: apps/explorer/lib/explorer/chain/transaction.ex (hunk 3 above)
Comment (zachdaniel) [reply apps/explorer/lib/explorer/chain/transaction.ex]: Yeah, I'll add some now.

**Pull request merged**